pkg/util/validation/limits.go

```diff
@@ -762,6 +768,10 @@ func (o *Overrides) ForwardingEndpoint(user string) string {
 	return o.getOverridesForUser(user).ForwardingEndpoint
 }
 
+func (o *Overrides) EphemeralChecker(user string) ephemeral.SeriesChecker {
+	return o.getOverridesForUser(user).EphemeralSeriesMatchers
```

Comment:
Since `EphemeralSeriesMatchers` is declared as `EphemeralSeriesMatchers ephemeral.LabelMatchers` and `ephemeral.LabelMatchers` is a struct, this will always be a non-nil value, even if user has no ephemeral matchers defined.

Switching to `EphemeralSeriesMatchers *ephemeral.LabelMatchers` needs to be handled carefully: we want to avoid returning non-nil interface value with `nil` pointer in it.

Reply:
Good point. I *think* this should always return a non-nil pointer now, because as you said the `EphemeralSeriesMatchers` property is a struct so it should always be instantiated. https://github.com/grafana/mimir/blob/386697701607e152cee94983d9c57305d61ee02b/pkg/util/validation/limits.go#L770-L772

Reply:
I think we should return `nil` in case user has no ephemeral matchers, to avoid the logic in distributor.

Reply:
What do you think about doing it this way: 2a6013feb

This `HasMatchers()` method is generic enough for it to also be implemented by other implementations of the `SeriesChecker`. That way we don't need to worry about returning an interface-typed `nil`.

Reply:
Adding `HasMatchers` seems fine to me, but distributor still needs to handle `nil` case gracefully to avoid crashes.

Reply:
I was trying to avoid using `reflect` in this hot path. Now it is guaranteed that the ephemeral checker is never `nil` (I should update the comment). Otherwise, to be sure that we never panic due to a typed `nil` interface, we'd have to check the interface value with `reflect`, no?

Reply:
Oh, I wouldn't do anything like that. In distributor, I'd just for nil interface value:

```
		ephemeralChecker := d.EphemeralCheckerByUser.EphemeralChecker(userID)
 		if ephemeralChecker != nil && ephemeralChecker.HasMatchers() {
```

In overrides, I'd return `nil` like this:
```
 func (o *Overrides) EphemeralChecker(user string) ephemeral.SeriesChecker {
 	m := o.getOverridesForUser(user).EphemeralSeriesMatchers
 	if len(m.config) > 0 {
 	 	return m
 	}
 	return nil
 }
```

That last return is fine, as we're returning `nil` interface value.

Reply:
Thx for the suggestion, that makes sense.

I removed `HasMatchers()` from the `ephemeral.SeriesChecker` interface.
I still left `HasMatchers()` in the `ephemeral.LabelsMatcher` implementation because from the `validation` package I can't access the private `m.config` property without exporting it, I figured that this config is quite an internal property so it would be better to just call `m.HasMatchers()` from `Overrides.EphemeralChecker()`:

https://github.com/grafana/mimir/blob/547b5f401b24750008704e88e78e8f7cbf5b790a/pkg/util/validation/limits.go#L770-L777 

Commit: 547b5f401